tests/scripts/branch-history

```diff
@@ -0,0 +1,151 @@
+
+   This script checks if the current branch history, defined as the commits that
+   will be merged in master (those shown in a GitHub PR), contains unusually
+   large files, unusually large number of changes in a commit, unusually large
+   number of commits, etc.
+
+   The specific checks are provided at 3 levels: branch, commit and file.
+
+   Branch-level checks:
+   - Check if the branch contains unusually large number of commits
+
+   Commit-level checks:
+   - Check if the commit changes an unusually large number of files
+   - Check if the commit changes an unusually large number of lines
+
+   File-level checks:
+   - Check if the file is binary
+   - Check if the file is unusually large
+   - Check if an unusually large number of lines have been added to the file
+   - Check if an unusually large number of lines have been deleted from the file
+
+   To be useful, the script should be run in a repository that is not in a
+   "detached HEAD" state.
+
+   Example usage: $0 ../..
+
+EOF
+      exit
+      ;;
+esac
+
+# Setup
+cur_dir=`pwd`
+mfem_dir=$1
+if [ "$mfem_dir" = "" ]; then
+    mfem_dir="../.."
+fi
+cd $mfem_dir
+mfem_dir=`pwd`
+shift
+
+# Maximum number of acceptable commits in the branch
+branch_max_commits=100
+# Maximum number of acceptable files changed in any commit in the branch
+commit_max_files_changed=100
+# Maximum number of acceptable lines changed in any commit in the branch
+commit_max_lines_changed=3000
+# Maximum acceptable size (in K) of any file changed in any commit in the branch
+file_max_size=200
+# Maximum number of acceptable lines added to any file changed in any commit in the branch
+file_max_lines_added=1000
+# Maximum number of acceptable lines deleted from any file changed in any commit in the branch
+file_max_lines_deleted=1000
+
+current_branch="$(git symbolic-ref --short HEAD)"
+
+# Check if the branch contains unusually large number of commits
+num_commits="$(git rev-list --count master..${current_branch})"
+if [ ${num_commits} -gt ${branch_max_commits} ]
+then
+    printf '\033[35m'"This branch contains unusually many commits: " 1>&2
+    printf "${num_commits} commits not in master"'\033[0m\n' 1>&2
+fi
+
+# Check all the commits in the current branch (i.e. the commits that will be
+# merged in master, as shown in a GitHub PR):
+while read -r rev; do
+    echo "Processing commit ${rev}..."
+    shortrev="$(git log -1 --format=%h ${rev})" # short commit hash
+
+    # Check if the commit changes an unusually large number of files
+    files_changed="$(git show --pretty="" --shortstat ${rev} | awk '{print $1}')"
+    if [ ${files_changed} -gt ${commit_max_files_changed} ]
+    then
+        printf '\033[36m'"Commit ${shortrev} changes unusually many files: " 1>&2
+        printf "${files_changed} files changed"'\033[0m\n' 1>&2
+    fi
+
+    # Check if the commit changes an unusually large number of lines
+    lines_changed="$(git show --pretty="" --shortstat ${rev} | awk '{print $4+$6}')"
```

Comment:
This seems to work just fine even if there are no insertions or deletions...

Reply:
So be it.

Reply:
See https://travis-ci.org/tzanio/mfem/jobs/639714032